### ISSUE TITLE
Solving all EXWM/Emacs and CUA standard key shortcuts conflicts once and for all

### ISSUE THREAD

**Issue by StacktraceException** [ch11ng/exwm#718]
Hello, the Super key is not used in Emacs because it is meant for Window Managers, but this restriction is unneeded because Window Management is possible without this key (like with C-x).

To make CUA shortcut keys (like C-v for pasting) and ALL Emacs keybindings work without conflicting nor switching modes nor using C-c C-q — it would be best to just globally change the commands to use Super instead of Ctrl (and use an additional keyboard key or Super+Ctrl instead of Alt).  
and turn off the EXWM Input Prefix Keys. The s-w and other workspace keybindings should be bound to C-x 6 in the logic of C-x 4 and C-x 5. (Organizing all UI with C-x 7 (and so on for hierarchy) combined with tags and search (like in org-mode) would be interesting?).

TLDR: How to change all the keybindings in Emacs global keymap to use Super key instead of Ctrl key?

**Comment by ch11ng** [comment]
With the method described in https://github.com/ch11ng/exwm/issues/635#issuecomment-549135323 , it's now possible to create mapping between prefix keys.  Let me know if it's not what you requested.

**Comment by StacktraceException** [comment]
By guesswork I managed to get this:


(setq exwm-manage-configurations '((t char-mode t)))

(defmacro exwm--with-gensyms (symbols &rest body)
  "Same as `org-with-gensyms.'"
  (declare (debug (sexp body)) (indent 1))
  `(let ,(mapcar (lambda (s)
  `(,s (make-symbol (concat "--" (symbol-name ',s)))))
                 symbols)
     ,@body))

(defmacro exwm--invoke-map (kbd)
  `(eval ,(exwm--with-gensyms (evlist keydesc)
   `(let* ((,evlist (listify-key-sequence ,kbd))
   (,keydesc (key-description ,evlist)))
      `(defun ,(intern (format "exwm--invoke-%s-map" ,keydesc)) ()
 ,(format "Invoke `%s' map." ,keydesc)
 (interactive)
 ,@(cl-loop for event in ,evlist
    append `((exwm-input--cache-event ,event t)
     (exwm-input--unread-event ,event))))))))

(customize-set-variable
 'exwm-input-global-keys
 (append exwm-input-global-keys
(cl-loop for (from-keys . to-keys) in '(("C-x" . "s-x")
("C-h" . "s-h")
("C-g" . "s-g")
("C-a" . "s-a")
("C-e" . "s-e")
("C-t" . "s-t")
("C-h" . "s-h")
("C-f" . "s-f")
("C-b" . "s-b")
("M-x" . "s-x")
("M-:" . "s-:")
("M-;" . "s-;")
("C-p" . "s-p")
("C-n" . "s-n")
("C-o" . "s-o")
("C-d" . "s-d")
("C-k" . "s-k")
("C-j" . "s-j")
("C-y" . "s-y")
("C-w" . "s-w")
("M-y" . "s-Y")
("C-SPC" . "s-SPC")
;; C-SPC doesn't work. Hyper keys (H-x for example) doesn't either. (Would
be much more ergonomic to use right ctrl as hyper). s-x s-s etc doesn't
work, because after s-x, the s-s gets interpreted as separate command.
("C-/" . "s-/")
("C-?" . "s-?")
("C-x b" . "s-z")
("C-x 1" . "s-1")
("C-x 2" . "s-2")
("C-x 3" . "s-3")
("C-x o" . "s-<f1>")
("C-x k" . "s-<f4>")
)
 collect (cons (vconcat (listify-key-sequence (kbd to-keys)))
(exwm--invoke-map (kbd from-keys))))))

(With this C-j C-t etc don't work in chromium, the minibuffer says Buffer
is read only)


The other code snippets require me to define the exwm-input-invoke-factory
from what I understand because of "exwm-input-invoke-factory function
definition is void.

Thank You for your expertise, from a non-programmer hobbyist. I actually
gave up until now, because I realized one can only get so far without
learning programming. If you would please send me anything better I would
be very happy. :)

**Comment by ch11ng** [comment]
Please refer to 'Usage Note 2' in https://github.com/ch11ng/exwm/issues/635#issuecomment-549365025.  For instance, with the following snippet, whenever <kbd>s-x</kbd> is pressed, Emacs would behave as if it receives <kbd>C-x</kbd> instead.  You only need to extend the list (`exwm-input-global-keys`).

```elisp
(exwm-input-invoke-factory "C-x")

(custom-set-variables
 '(exwm-input-global-keys
   '(([?\s-x]
      . exwm-input--invoke--C-x))))
```

**Comment by StacktraceException** [comment]
I'm appreciative of your feedback, and my previous long failures ar
understanding E?Lisp code had discouraged me from trying more, although now
after more guesswork I'm stuck on getting errors:

definition is void (about the (exwm-input-invoke-factory "C-x") or
something like cannot invoke c-x--map (when "M-x" . "s-x") is
commented out. (Was trying to see if it would help).



This is one of the things I tried:

;;(exwm-input-invoke-factory "C-x" ) ; exwm-input--invoke--C-x
;;(exwm-input-invoke-factory "C-x")

(custom-set-variables
 '(exwm-input-global-keys
   '(([?\s-x]
      . exwm-input--invoke--C-x))))
;; (exwm-input-invoke-factory "C-h" ) ; exwm-input--invoke--C-h
;; (exwm-input-invoke-factory "M-x" ) ; exwm-input--invoke--M-x
;; (exwm-input-invoke-factory "M-:" ) ; exwm-input--invoke--M-:
;; (exwm-input-invoke-factory "C-x w" ) ; exwm-input--invoke--C-x\ w


(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(custom-safe-themes
   '("a27c00821ccfd5a78b01e4f35dc056706dd9ede09a8b90c6955ae6a390eb1c1e"
"3c83b3676d796422704082049fc38b6966bcad960f896669dfc21a7a37a748fa"
"a27c00821ccfd5a78b01e4f35dc056706dd9ede09a8b90c6955ae6a390eb1c1e"))
 '(display-time-24hr-format t)
 '(display-time-day-and-date t)
 '(enable-recursive-minibuffers t)
 '(exwm-input-global-keys
   '(([?\s-x]
      . exwm-input--invoke--C-x)
([8388722]
 . exwm-reset)
([8388727]
 . exwm-workspace-switch)
([8388646]
 lambda

and so on

I also tried to use your snippet outside of custom-set-variables like this:

(exwm-input-invoke-factory "C-x")

(custom-set-variables
 '(exwm-input-global-keys
   '(([?\s-x]
      . exwm-input--invoke--C-x))))

I  try different combinations until things work, because I don't
understand much, sorry —  I could understand that I could be annoying
to still not understand…


niedz., 12 kwi 2020 o 14:18 Chris Feng <notifications@github.com>
napisał(a):

**Comment by ch11ng** [comment]
Sorry but I forgot to mention that the feature (`exwm-input-invoke-factory`) has not been released yet.
You'll have to install EXWM from source at the moment.  Please refer to the [steps](https://github.com/ch11ng/exwm/wiki#install-from-source) in the wiki for how to do this.

**Comment by StacktraceException** [comment]
Thanks :). I installed from source, now there is no "definition is void"
error, but after more guesswork I got only thus far:

C-S-x from the  comment you linked works but only in Emacs' buffers (not
chromium's).
Interestingly it also works with chaining such as C-S-x S-M-x, but not with
combining s-x for example. (I get C-x s-f is undefined for example).

I tried disabling the previous code like before to see if there were any
conflicts, but just got that cant invoke -- map error or something like
that again like previously.


wt., 5 maj 2020 o 15:27 Chris Feng <notifications@github.com> napisał(a):

**Comment by ch11ng** [comment]
The original post defined rules such that whenever you pressed <kbd>C-S-x</kbd> *as the prefix key*, you got <kbd>C-x</kbd> instead.
So you'll notice that <kbd>C-S-x S-M-x</kbd> works if <kbd>C-x S-M-x</kbd> happens to be defined, and with <kbd>C-x s-f</kbd> undefined, <kbd>C-S-x s-f</kbd> won't work.
Perhaps you just need to tidy up your configurations a bit and try the example in my previous comment (https://github.com/ch11ng/exwm/issues/718#issuecomment-612605731), and see if you can replace <kbd>C-x</kbd> with <kbd>s-x</kbd> as the prefix key.
